src/main/java/com/Elessar/app/server/LogOnHandler.java

```diff
@@ -36,24 +45,30 @@ public void handle(HttpExchange he) throws IOException {
             final LogonRequest logonRequest = LogonRequest.parseFrom(is);
             final LogonResponse.Builder logonResponse = LogonResponse.newBuilder();
             final String userName = logonRequest.getName();
-            final String password = logonRequest.getPassword();
-            if (!userData.containsKey(userName)) {
-                logonResponse.setSuccess(false).setFailReason("User Name " + userName + " is NOT Registered !");
-                he.sendResponseHeaders(400, 0);
-            } else if (!userData.get(userName).getPassword().equals(password)){
-                //System.out.printf("user = %s, password1 = %s, password2 = %s\n", userName, userData.get(userName).getPassword(), password);
-                logonResponse.setSuccess(false).setFailReason("Incorrect Password !");
-                he.sendResponseHeaders(400, 0);
-            } else if (userData.get(userName).getOnlineStatus()) {
-                logonResponse.setSuccess(false).setFailReason(userName + " is already log on !");
+            final Map<String, String> filters = new HashMap<>();
+            filters.put("name", logonRequest.getName());
+            filters.put("password", logonRequest.getPassword());
+
+            Iterator<Document> cursor = users.find(filters).iterator();
```

Comment:
What you are doing here requires 2 database operations: first query then update (if found).
MongoDB has methods such as findOneAndReplace() and findOneAndUpdate(), which can be more efficient.

Similar can be done for logoff as well.